src/megaclient.cpp

```diff
@@ -919,26 +919,58 @@ void MegaClient::activateoverquota(dstime timeleft)
                     t->state = TRANSFERSTATE_RETRYING;
                     t->slot->retrybt.backoff(timeleft);
                     t->slot->retrying = true;
-                    app->transfer_failed(t, API_EOVERQUOTA, timeleft);
+                    app->transfer_failed(t, API_EOVERQUOTA, timeleft, targetHandle);
                     ++performanceStats.transferTempErrors;
                 }
             }
         }
     }
-    else if (setstoragestatus(STORAGE_RED))
+    else    // Storage overquota
     {
-        LOG_warn << "Storage overquota";
-        for (transfer_map::iterator it = transfers[PUT].begin(); it != transfers[PUT].end(); it++)
+        if (isForeignNode(targetHandle))
         {
-            Transfer *t = it->second;
-            t->bt.backoff(NEVER);
-            if (t->slot)
+            // If target is foreign, remove it from all transfers
+            for (transfer_map::iterator it = transfers[PUT].begin(); it != transfers[PUT].end(); it++)
             {
-                t->state = TRANSFERSTATE_RETRYING;
-                t->slot->retrybt.backoff(NEVER);
-                t->slot->retrying = true;
-                app->transfer_failed(t, API_EOVERQUOTA, 0);
-                ++performanceStats.transferTempErrors;
+                Transfer *t = it->second;
+                for (file_list::iterator it = t->files.begin(); it != t->files.end();)
+                {
+                   File *f = (*it++);
+                   if (f->h == targetHandle)
+                   {
+                       app->transfer_failed(t, API_EOVERQUOTA, 0, targetHandle);
+                       t->removeTransferFile(API_EOVERQUOTA, f, mTctableRequestCommitter);
+                       continue;
+
+                   }
+                   if (t->files.empty())
+                   {
+                       LOG_debug << "Removing transfer";
+                       t->state = TRANSFERSTATE_FAILED;
+                       t->finished = true;
+                       app->transfer_removed(t);
+                       ++performanceStats.transferTempErrors;
+                       delete t;
+                   }
+                }
+            }
+            return;
+        }
+        else
```

Comment:
If a transfer with private target results in a call to `activateoverquota()` before a call due to a foreign target, all transfers will fail (`transfer_failed()`), including any with a foreign target.